### PR TITLE
ENH add demsar's zero_method strategy to the wilcoxon test

### LINKED ISSUE #17530: Unnecessary approximation in `scipy.stats.wilcoxon(x, y)`
Context: `scipy.stats.wilcoxon(x, y)` -- `x` and `y` being 1D arrays of same size -- has two modes, `approx` and `exact`, and three policies for treating ties (i.e. when `x[i] == y[i]`), of which I believe there is an issue when `zero_method="zsplit"`.

If there is at least one tie, the `mode` is switched to `approx` independently of the chosen `zero_method`. 

This happens on this line: [https://github.com/scipy/scipy/blob/de80faf9d3480b9dbb9b888568b64499e0e70c19/scipy/stats/_morestats.py#L3335](https://github.com/scipy/scipy/blob/de80faf9d3480b9dbb9b888568b64499e0e70c19/scipy/stats/_morestats.py#L3335).

Based on [1], I believe it shouldn't be necessary to use the `approx` method with `zero_method="zsplit"`.

In other words, the gaussian approximation, which normally should be used with `x.size > 50`, is causing unnecessary imprecision in the test computation. 

[1] J. Demšar, “Statistical Comparisons of Classifiers over Multiple Data Sets,” Journal of Machine Learning Research, vol. 7, no. 1, pp. 1–30, 2006. Url: http://jmlr.org/papers/v7/demsar06a.html

---

Paragraph extracted from [1] (section "3.1.3 WILCOXON SIGNED-RANKS TEST", page 7):

> Let di again be the difference between the performance scores of the two classifiers on i-th out of N data sets. The differences are ranked according to their absolute values; **average ranks are assigned in case of ties**. Let R+ be the sum of ranks for the data sets on which the second algorithm outperformed the first, and R− the sum of ranks for the opposite. Ranks of di = 0 are split evenly among the sums; **if there is an odd number of them, one is ignored**:

--- 

I already made a (dirty) fix with the recommendations above on a local copy so I can open a (to-be-cleaned) PR, but I'd apreciate some validation from the maintainers that this makes sense before going through the review process.

### PR DESCRIPTION
**Pull request by jpcbertoldo**
#### Reference issue

Closes https://github.com/scipy/scipy/issues/17530 

#### What does this implement/fix?

A better strategy to deal with cases where there are zeros entries (in ``d``) in the wilcoxon signed rank test.

#### Additional information

See https://gist.github.com/jpcbertoldo/da22764eed915ee1f8a1d79e32aa72f6?permalink_comment_id=4423840#gistcomment-4423840

### CODE REVIEW

**Comment by jpcbertoldo** [comment]
I don't understand if shoudl write `.. versionadded:: 1.11.0` here.

File: scipy/stats/_morestats.py [hunk 1]
```diff
@@ -3512,9 +3518,26 @@ def wilcoxon(x, y=None, zero_method="wilcox", correction=False,
...
+                d = np.delete(d, where_idx)
+                n_zero -= 1
+                
+        else:
+            mode = "approx"
+            warnings.warn("Exact p-value calculation does not work if there are "
+                        "zeros. Switching to normal approximation.")
 
     if mode == "approx":
         if zero_method in ["wilcox", "pratt"]:
```
Comment (jpcbertoldo) [inline scipy/stats/_morestats.py:3543]: Probably for another PR, but it feels like all these `if`s became too complex.

Perhaps an exhaustive `if ... and ... and ...` with the values of `method`, `zero_method` would be easier to understand?

Comment (mdhaber) [reply scipy/stats/_morestats.py:3543]: It is getting complex, but let's not refactor that right now.

**Comment by mdhaber** [comment]
Most of the CI failures are not your fault, but please fix the PEP8 issues that are causing scipy.scipy (Main Lint) to fail:
```
scipy/stats/_morestats.py:3421:61: W291 trailing whitespace
scipy/stats/_morestats.py:3422:75: W291 trailing whitespace
scipy/stats/_morestats.py:3424:50: W291 trailing whitespace
scipy/stats/_morestats.py:3439:67: W291 trailing whitespace
scipy/stats/_morestats.py:3440:73: W291 trailing whitespace
scipy/stats/_morestats.py:3441:76: W291 trailing whitespace
scipy/stats/_morestats.py:3521:1: W293 blank line contains whitespace
scipy/stats/_morestats.py:3523:1: W293 blank line contains whitespace
scipy/stats/_morestats.py:3528:67: W291 trailing whitespace
scipy/stats/_morestats.py:3536:1: W293 blank line contains whitespace
scipy/stats/_morestats.py:3539:80: E501 line too long (81 > 79 characters)
scipy/stats/_morestats.py:3540:25: E128 continuation line under-indented for visual indent
```

File: scipy/stats/_morestats.py [hunk 2]
```diff
@@ -3512,9 +3518,26 @@ def wilcoxon(x, y=None, zero_method="wilcox", correction=False,
...
+            # ties are accepted but the number of zeros must be even
+            # if it is odd, one zero is discarded
+            if n_zero % 2 == 1:
+                # discard one zero difference
+                # first [0] is for the tuple returned by np.where, 
+                # second [0] is for the first element of the tuple
+                where_idx = np.where(d == 0)[0][0]
+                x = np.delete(x, where_idx)
+                if y is not None:
+                    y = np.delete(y, where_idx)
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/stats/_morestats.py:3533]: The paper says to compute ranks first, before the zero is discarded. I think this matters.

If I'm reading correctly, then, you don't even need to remove the zero from `d`. The change up here would just be:
```
    if n_zero > 0 and mode == "exact":
```
changes to:
```
    if n_zero > 0 and mode == "exact" and zero_method != 'zsplit':
```

Then all you need to do down below where you see
```python3
if zero_method == "zsplit":
```
is adjust the calculation of `rzero`. It's the count of the number of zeros rounded down to a multiple of two multiplied by the rank that is assigned to zeros.

**Review by mdhaber — CHANGES_REQUESTED** [review]

**Comment by mdhaber** [comment]
Please also add tests with both even and odd numbers of zeros. Ideally, you would check the p-value against an external implementation that uses this method.

**Comment by mdhaber** [comment]
@jpcbertoldo Just a reminder that you have a review waiting. If you have questions, let me know.

**Comment by jpcbertoldo** [comment]
Hey @mdhaber , I lost track of this since a while (vacation, other priorities...). I will try to come back to it soon!

**Comment by lucascolley** [comment]
gentle reminder @jpcbertoldo :)

**Comment by mdhaber** [comment]
I am not sure this is needed anymore now that the permutation test is available. Did you want to review it @lucascolley?

**Comment by lucascolley** [comment]
Ah yes, I missed your PR to close the linked issue - feel free to close!

**Pull request closed without merging**